Performance improvement of unit tests: Remove delays and inject dependencies

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11228
Context: pdfdoF-64g-p2#comment-7238
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This is a continuation of https://github.com/woocommerce/woocommerce-ios/pull/14802 to investigate the reasons of slow unit tests and look for solutions.

I haven't found more all-encompassing widely applied solutions. Therefore this PR has a moderate average speed execution improvement [from around 50ms to 40ms.](https://buildkite.com/organizations/automattic/analytics/suites/woocommerce-ios/runs/5f76f05d-a5bc-86ec-a3dd-93affa86069d).

### Jetpack Setup ViewModel Enhancements:
* Added a new property `delayBeforeRetry` to the `JetpackSetupViewModel` and updated its initialization to include this property. This change allows for configurable retry delays during Jetpack setup. [[1]](diffhunk://#diff-5b2d3d7f53b79702e24015df8d058f2c45f0b6ba5e79eba6db48da378af75cccR99-R105) [[2]](diffhunk://#diff-5b2d3d7f53b79702e24015df8d058f2c45f0b6ba5e79eba6db48da378af75cccR114) [[3]](diffhunk://#diff-5b2d3d7f53b79702e24015df8d058f2c45f0b6ba5e79eba6db48da378af75cccL328-R331) [[4]](diffhunk://#diff-5b2d3d7f53b79702e24015df8d058f2c45f0b6ba5e79eba6db48da378af75cccL343-R346)
* Updated `JetpackSetupViewModelTests` to include the new `delayBeforeRetry` parameter in the test cases. [[1]](diffhunk://#diff-1da69198084f92cd1adae036140fa76a6306ed8f38155afc5beb212b61003888L19-R35) [[2]](diffhunk://#diff-1da69198084f92cd1adae036140fa76a6306ed8f38155afc5beb212b61003888L45-R45) [[3]](diffhunk://#diff-1da69198084f92cd1adae036140fa76a6306ed8f38155afc5beb212b61003888L71-R71) [[4]](diffhunk://#diff-1da69198084f92cd1adae036140fa76a6306ed8f38155afc5beb212b61003888L92-R92) [[5]](diffhunk://#diff-1da69198084f92cd1adae036140fa76a6306ed8f38155afc5beb212b61003888L116-R116) [[6]](diffhunk://#diff-1da69198084f92cd1adae036140fa76a6306ed8f38155afc5beb212b61003888L141-R141)

### Push Notification Background Synchronizer Refactoring:
* Introduced `PushNotificationBackgroundSynchronizerProtocol` and updated `PushNotificationBackgroundSynchronizer` to conform to this protocol.
* Created `PushNotificationBackgroundSynchronizerFactory` and its corresponding protocol to handle the creation of synchronizer instances.
* Updated `PushNotificationsManager` to use the new factory for creating background synchronizers. [[1]](diffhunk://#diff-9e8b41ce9b767f7a2ef32434891f864524459ee3434d14b1573b3f4858f15d61R115-R125) [[2]](diffhunk://#diff-9e8b41ce9b767f7a2ef32434891f864524459ee3434d14b1573b3f4858f15d61L318-R322)

### Hub Menu ViewModel Enhancements:
* Refactored `HubMenuViewModel` to make `viewDidAppear`, `refreshGoogleAdsCampaignCheck`, and `refreshBlazeEligibilityCheck` asynchronous functions. Added a helper extension to handle these asynchronous calls. [[1]](diffhunk://#diff-a7353a3c9b54e18ce66e192000f034059a624c6824a359575150dc1788697d04L173-R188) [[2]](diffhunk://#diff-a7353a3c9b54e18ce66e192000f034059a624c6824a359575150dc1788697d04L210-L223) [[3]](diffhunk://#diff-a7353a3c9b54e18ce66e192000f034059a624c6824a359575150dc1788697d04R484-R504)

### Remote Order Synchronizer Enhancements:
* Added a `debounceDuration` property to `RemoteOrderSynchronizer` to allow configurable debounce intervals for sync requests. [[1]](diffhunk://#diff-0667e76ef83e3a4ed8c121e825fa2a1fcc649011939b3755888c8216d1fa8c0eR88-R104) [[2]](diffhunk://#diff-0667e76ef83e3a4ed8c121e825fa2a1fcc649011939b3755888c8216d1fa8c0eL416-R423)

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

⚠️  There has been flakiness with  `ConfigurableBundleProductViewModelTests` `test_bundleItemViewModels_have_correct_quantity_when_parent_order_item_quantity_is_more_than_one` in CI runs that I cannot quite figure out. It runs super-fast locally but slows down significantly on the CI. I will make sure this branch is stable and has a few consecutive successful runs on CI before merging.


## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Always inject MockStorageManager in OrderListViewModelTests

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Make FirstProductCreatedViewModelTests synchronous to avoid costly isInverted checks

staskus committed:
Limit reliance on using timeouts in HubMenuViewModelTests

Make viewDidAppear of HubMenuViewModel async which allows the tests to avoid timeouts and more importantly isInverted calls to assert that somehow is not called. It improves the speed of the tests dramatically.

staskus committed:
Set debounceDuration for RemoteOrderSynchronizer to speed up tests

staskus committed:
Update OrderListViewModelTests.swift

staskus committed:
Remove unnecessary async usage in CollectOrderPaymentUseCaseTests tests

staskus committed:
Speed up GoogleAdsCampaignReportCardViewModelTests tests by always using mocked store response

Some tests are slowed down since they rely on a real store to retrieve data

staskus committed:
Update ConfigurableBundleProductViewModelTests.swift

staskus committed:
Mock backgroundSynchronizer in PushNotificationsManagerTests for faster execution

staskus committed:
Remove waitUntil from ConfigurableBundleProductViewModelTests tests

Use Combine instead to remove ambiguity and only wait for validation as long as needed

staskus committed:
Remove delays from JetpackSetupViewModelTests

staskus committed:
Update ConfigurableBundleProductViewModelTests.swift

staskus committed:
Merge branch 'trunk' into fix/11228-performance-improvement-of-slow-tests-2

staskus committed:
Update ProductFormRemoteActionUseCaseTests.swift

staskus committed:
Merge branch 'trunk' into fix/11228-performance-improvement-of-slow-tests-2

staskus committed:
Wait separately for conditions to be satisfied in MainTabBarControllerTests

staskus committed:
Inject zero delays to ShippingLabelStoreTests

staskus committed:
Inject zero delays to WooShippingStoreTests

staskus committed:
Address ConfigurableBundleProductViewModelTests flakiness by making sure all the tests complete resource loading

staskus committed:
Merge branch 'trunk' into fix/11228-performance-improvement-of-slow-tests-2

staskus committed:
Increase stability of BlazeLocalNotificationSchedulerTests

staskus committed:
Revert changes to ConfigurableBundleProductViewModelTests due to ineffectiveness

staskus committed:
Set debounce lower than timeout for RemoteOrderSynchronizerTests

staskus committed:
Increase timeout for ConfigurableBundleProductViewModelTests

staskus committed:
Merge branch 'trunk' into fix/11228-performance-improvement-of-slow-tests-2

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swift:
```diff
@@ -477,6 +481,27 @@ private extension HubMenuViewModel {
 
+// MARK: - Helpers
+extension HubMenuViewModel {
+    func viewDidAppear() {
+        Task { @MainActor in
+            await viewDidAppear()
+        }
+    }
```
Oh! That's an interesting one 🤔

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewModels/CardPresentPayments/CollectOrderPaymentUseCaseTests.swift:
```diff
@@ -77,7 +77,7 @@ final class CollectOrderPaymentUseCaseTests: XCTestCase {
     }
 
-    func test_collectPayment_processing_completion_tracks_payment_success_event() async throws {
+    func test_collectPayment_processing_completion_tracks_payment_success_event() throws {
```
Do you know why would the test be slower just because (miss)uses the `async` keyword? There's no await inside so it shouldn't be any different than a non-async function, no?

staskus replied:
I'm not sure, there may be no difference. There's some weirdness and inconsistency around these tests and the results differ between my local machine and CI. Either way, since `async` is not used we can remove it.

iamgabrielma commented on WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/JetpackSetupViewModel.swift:
```diff
@@ -96,11 +96,13 @@ final class JetpackSetupViewModel: ObservableObject {
          connectionOnly: Bool,
          stores: StoresManager = ServiceLocator.stores,
          analytics: Analytics = ServiceLocator.analytics,
+         delayBeforeRetry: Double = Constants.delayBeforeRetry,
```
Simple but effective 💯

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductBundles/ConfigurableBundleProductViewModelTests.swift:
```diff
@@ -341,7 +341,8 @@ final class ConfigurableBundleProductViewModelTests: XCTestCase {
         // The products are loaded async before the bundle item view models are set.
-        waitUntil {
+        // This test occasionally fails on CI due to slowness, cannot reproduce locally.
+        waitUntil(timeout: 10.0) {
```
😨 it passes locally with a timeout of 1 or less, and I cannot see any reason right away of why would fail in CI. 

Just wondering here if the session we're injecting in the mock would have any effect into this. We're using `.makeForTesting()` while in other tests we use`.testingInstance` . I'd say no, but also we don't need the whole session details to init this class.

Outside of this PR and as a side note, I don't think `makeForTesting` name is very good, there's no clear difference with `.testingInstance`

staskus replied:
I was banging my head with this test. I haven't tried playing with `testingInstance` vs `makeForTesting`, I probably should. Although other tests in this class rely on it as well. 

This works fast on local machine and incredibly slow on the CI...

staskus replied:
I tried to make this change just for experiment purpose and will check tomorrow.

staskus replied:
It still runs in over 5 seconds... It's a true mystery to me. 

<img width="900" alt="image" src="https://github.com/user-attachments/assets/e5ed4839-0d3c-4e76-b78d-ab4b405237b6" />

staskus replied:
The performance improvements are still solid. I'm gonna merge.

iamgabrielma reviewed (approved):
LGTM 🚀

staskus committed:
Use testingInstance in ConfigurableBundleProductViewModelTests
